### PR TITLE
[releases] Automating releases

### LINKED ISSUE #76: [release] automating releases
The `make publish` command helps me to put the docker images on docker hub, but that's not enough, we want use `Makefile` to be able:

1. Tag the new release with a version number, example `openwisp/openwisp-base:0.1.0` (Images with `latest` tag should also be uploaded.)
2. User to easily upload images in their own docker hub account, example `atb00ker/openwisp-base:0.1.0`
3. ~Release edge image directly from the master branch as `openwisp/openwisp-base:edge`~

Edit:
- 3rd point taken care by docker hub autobuilds! :smile:

### PR DESCRIPTION
**Pull request by marfgold1**
This will allow to automate releases as described in the issue
Closes #76

### CODE REVIEW

File: build/openwisp_base/Dockerfile [hunk 1]
```diff
@@ -20,12 +20,12 @@ ENV PYTHONPATH=/install/lib/python3.7/site-packages
 # Python Packages
 RUN pip install --editable git+git://github.com/openwisp/openwisp-controller#egg=openwisp_controller \
                 --prefix='/install'
+RUN pip install --upgrade https://github.com/openwisp/django-freeradius/tarball/master \
```
Comment (2stacks) [inline build/openwisp_base/Dockerfile]: Was this change intentional?  Just curious if django-freeradius needed to be installed before openwisp-radius.

File: .travis.yml [hunk 2]
```diff
@@ -13,4 +13,4 @@ jobs:
   include:
     - stage: tests
       script:
-      - make travis-runtests
+      - make publish-travis
```
Comment (2stacks) [inline .travis.yml]: Update docs for this change from travis-runtests to publish-travis.

File: build/openwisp_base/Dockerfile (hunk 1 above)
Comment (marfgold1) [reply build/openwisp_base/Dockerfile]: Yeah it's an [ongoing issue](https://github.com/openwisp/openwisp-radius/issues/73), just a temporary fix.

File: .travis.yml (hunk 2 above)
Comment (marfgold1) [reply .travis.yml]: Okay, I'll do it with other part (this is still not done btw)

**Comment by marfgold1** [comment]
I am done, and it does work, but I'm unsure if this is the correct way to do it :sweat_smile: 
My custom images can be found here:
https://github.com/marfgold1/docker-openwisp/packages
https://hub.docker.com/u/marfgold1
My successful fork build (with uploading the image too, of course):
https://travis-ci.org/marfgold1/docker-openwisp/builds/630898164
You need to set `DOCKER_USERNAME`, `DOCKER_PASSWORD`, `GITHUB_USERNAME`, and `GITHUB_TOKEN` in the travis dashboard. `GITHUB_TOKEN` is a personal access token to the your github, but you can use your github password (though I'm not recommend it)

**Comment by 2stacks** [comment]
I have some design related questions that you can probably address on Gitter (https://gitter.im/openwisp/dockerize-openwisp)  I don't want to fill up the thread here.

File: build/openwisp_base/Dockerfile [hunk 3]
```diff
@@ -4,7 +4,7 @@ FROM python:3.7.5-slim-buster AS SYSTEM
 # 1. Pillow: zlib1g-dev libjpeg-dev
 # 2. gettext: Required by envsubst used in scripts
 RUN apt update && \
-    apt install --yes --no-install-recommends zlib1g-dev libjpeg-dev gdal-bin gettext && \
+    apt install --yes --no-install-recommends zlib1g-dev libjpeg-dev gdal-bin gettext libcairo2 libpango-1.0-0 libpangocairo-1.0-0 && \
```
Comment (atb00ker) [inline build/openwisp_base/Dockerfile]: Also, change this as followed:

```
# System requirements:
# 1. Pillow: zlib1g-dev libjpeg-dev
# 2. gettext: Required by envsubst used in scripts
# 3. django-freeradius/weasyprint: libcairo2 libpangocairo-1.0
RUN apt update && \
    apt install --yes --no-install-recommends zlib1g-dev libjpeg-dev \
                libcairo2 libpangocairo-1.0 gdal-bin gettext && \
    rm -rf /var/lib/apt/lists/* /root/.cache/pip/* /tmp/*
```

File: Makefile [hunk 4]
```diff
@@ -46,16 +49,47 @@ develop: publish-build
 	docker-compose logs -f
 
 # Clean
+clean: TAG := latest
 clean:
```
[the remark this answers is not in the document]
Comment (marfgold1) [reply Makefile]: Yeah it is a functionality, called target-specified variable. It will set variable when the target gets called.

File: Makefile [hunk 5]
```diff
@@ -46,16 +49,47 @@ develop: publish-build
...
-	docker rmi --force openwisp/openwisp-base:latest \
-				openwisp/openwisp-base:intermedia-system \
-				openwisp/openwisp-base:intermedia-python \
+	docker rmi --force $(DOCKER_USERNAME)/openwisp-base:$(TAG) \
+				$(DOCKER_USERNAME)/openwisp-base:intermedia-system \
+				$(DOCKER_USERNAME)/openwisp-base:intermedia-python \
+				`docker images -f "dangling=true" -q` || true
+
+release-clean: clean
+	docker-compose rm -svf
```
[the remark this answers is not in the document]
Comment (marfgold1) [reply Makefile]: Release-clean will purge images with `latest` and `VERSION` tag, whether clean will purge images with `latest` tag only.

File: Makefile [hunk 6]
```diff
@@ -46,16 +49,47 @@ develop: publish-build
...
 	docker-compose stop
 	docker-compose down --remove-orphans --volumes --rmi all
 	docker-compose rm -svf
-	docker rmi --force openwisp/openwisp-base:latest \
-				openwisp/openwisp-base:intermedia-system \
-				openwisp/openwisp-base:intermedia-python \
+	docker rmi --force $(DOCKER_USERNAME)/openwisp-base:$(TAG) \
+				$(DOCKER_USERNAME)/openwisp-base:intermedia-system \
+				$(DOCKER_USERNAME)/openwisp-base:intermedia-python \
+				`docker images -f "dangling=true" -q` || true
```
[the remark this answers is not in the document]
Comment (marfgold1) [reply Makefile]: Well dont blame me, It's already been there in the original, so... (check unchanged line 56 or modified line 67)

[the remark this answers is not in the document]
Comment (marfgold1) [reply Makefile]: I will see if its work without it

File: Makefile [hunk 7]
```diff
@@ -46,16 +49,47 @@ develop: publish-build
...
+release-clean: clean
+	docker-compose rm -svf
+	docker rmi --force $(DOCKER_USERNAME)/openwisp-base:$(TAG) \
+				$(DOCKER_USERNAME)/openwisp-base:intermedia-system \
+				$(DOCKER_USERNAME)/openwisp-base:intermedia-python \
 				`docker images -f "dangling=true" -q` || true
 
 # Publish
+publish: TAG := latest
 publish: publish-build develop-runtests
```
[the remark this answers is not in the document]
Comment (marfgold1) [reply Makefile]: It is a functionality...

File: Makefile [hunk 8]
```diff
@@ -46,16 +49,47 @@ develop: publish-build
...
-	docker push openwisp/openwisp-base:latest
+	docker login -u $(DOCKER_USERNAME)
```
[the remark this answers is not in the document]
Comment (marfgold1) [reply Makefile]: I do agree, but sometimes user just forgot if they were logged in or not. It's kind of silly if the makefile stops at the docker push and you have to login first, then call make again

[the remark this answers is not in the document]
Comment (marfgold1) [reply Makefile]: But yeah, I can remove it if you want :wink:

File: Makefile [hunk 9]
```diff
@@ -46,16 +49,47 @@ develop: publish-build
...
+	docker push $(DOCKER_USERNAME)/openwisp-base:$(TAG)
+	docker-compose push
+
+release-publish: publish
+	docker push $(DOCKER_USERNAME)/openwisp-base:$(TAG)
+	docker-compose push
+
+travis-publish: TAG := nightly
+travis-publish: travis-runtests
```
[the remark this answers is not in the document]
Comment (marfgold1) [reply Makefile]: Yes it is :+1:

File: Makefile [hunk 10]
```diff
@@ -46,16 +49,47 @@ develop: publish-build
...
+	echo "$(DOCKER_PASSWORD)" | docker login -u "$(DOCKER_USERNAME)" --password-stdin
+	docker push $(DOCKER_USERNAME)/openwisp-base:$(TAG)
 	docker-compose push
+	echo "$(GITHUB_TOKEN)" | docker login docker.pkg.github.com -u $(GITHUB_USERNAME) --password-stdin
+	docker tag $(DOCKER_USERNAME)/openwisp-base:$(TAG) docker.pkg.github.com/$(GITHUB_USERNAME)/docker-openwisp/openwisp-base:$(TAG)
+	docker push docker.pkg.github.com/$(GITHUB_USERNAME)/docker-openwisp/openwisp-base:$(TAG)
+	for img in $$(docker-compose config | awk '{if($$0 ~ /image: $(DOCKER_USERNAME)/ && !visited[$$0]++) print $$2}'); do \
```
[the remark this answers is not in the document]
Comment (marfgold1) [reply Makefile]: Do you mean readibility in the travis logs itself or the thing in the modified line 88 (for img in ...)?
If it's travis, yeah I do agree. Should I make the docker login and docker push output to `/dev/null`?
If it's about that modified line 88, well it's important because you have to tag all of those images again. I will think of another approach

File: Makefile [hunk 11]
```diff
@@ -46,16 +49,47 @@ develop: publish-build
...
+		github_docker_repo=`echo $$img | sed "s/$(DOCKER_USERNAME)\//docker.pkg.github.com\/$(GITHUB_USERNAME)\/docker-openwisp\//g"`; \
+    	docker tag $$img $$github_docker_repo; \
+		docker push $$github_docker_repo; \
+	done
+
+test:
```
[the remark this answers is not in the document]
Comment (marfgold1) [reply Makefile]: Yes it is a mistake, I forgot to remove this line :sweat_smile:

File: build/openwisp_base/Dockerfile (hunk 3 above)
Comment (atb00ker) [reply build/openwisp_base/Dockerfile]: I made https://github.com/openwisp/docker-openwisp/pull/79 because I had to fix another issue which was stopping another GCI task, I'll merge it after @2stacks reviews it and you can rebase to it! :smile:

File: README.md [hunk 12]
```diff
@@ -166,12 +168,17 @@ Now you'll need to do steps (2) everytime you make a changes and want to build t
 
 The Makefile has following options, that are useful for developers & users. Default action is `compose-build`.
 
+**NOTE:** When you are managing your custom images, please change `DOCKER_USERNAME` with your docker account by export environment variable or passing directly to the make command. For example, if you want to publish your image, use `make publish DOCKER_USERNAME=<your-docker-username>`
```
[the remark this answers is not in the document]
Comment (marfgold1) [reply README.md]: Well, should I just mentioned it in `Build (Development)` section? I'm afraid that users will not notice this because it is an important notice whether it's in Deployment, Build (Development), and even Makefile. If users want to deploy their custom images, then they should set `DOCKER_USERNAME` and it's appropriate tag according to their own image, then run `docker-compose up`. That is important and if the users forgot, then docker-compose will use openwisp latest images instead. Also I do agree with that example, but then if you want to use multiple `make` calls, you have to export the variable instead of repeating yourself adding that `DOCKER_USERNAME=*`, that's why I mentioned it in the Makefile section.
So which place do you think the better for this notice?

File: README.md [hunk 13]
```diff
@@ -166,12 +168,17 @@ Now you'll need to do steps (2) everytime you make a changes and want to build t
...
 - `compose-build`: Build the openwisp images. This option must be used in most usecases, example when a user is building custom images.
 - `publish-build`: Similar to `compose-build` except the produced can be used for manually testing before publishing official images and publishing the official images using the `publish` option.
 - `runtests`: Run testcases to ensure all the services are working.
 - `develop-runtests`: Similar to `runtests`, it runs the testcases except doesn't stop the containers after running the tests which maybe desired for debugging & analyzing failing container's logs.
 - `travis-runtests`: Similar to `runtests`, it runs the testcases except used in travis builds because it doesn't clean up after running the tests and prints all the container's logs when an error occurs in the tests.
-- `clean`: Purge everything produced during building the images.
+- `clean`: Purge everything produced during building the images (only images that have 'latest' tag will be purged).
+- `release-clean`: Similar to `clean`, except it will also purge images that have 'VERSION' tag.
```
[the remark this answers is not in the document]
Comment (marfgold1) [reply README.md]: I do think like you just said, but thats require a changing tag and repeat the lines again. I think we can use a little shell script there. I'll see...

File: README.md [hunk 14]
```diff
@@ -166,12 +168,17 @@ Now you'll need to do steps (2) everytime you make a changes and want to build t
...
 - `develop`: Useful for development. Bundles all the commands required to build the images and run containers during development. With this option, you can interact with the interface and APIs and the logs will be printed on the terminal for debugging on the debugging level you select from the `.env` file.
-- `publish`: Build, test and publish the latest Official images.
+- `publish`: Build, test and publish the latest Official images to docker hub.
```
[the remark this answers is not in the document]
Comment (marfgold1) [reply README.md]: Will do :smile:

File: README.md [hunk 15]
```diff
@@ -166,12 +168,17 @@ Now you'll need to do steps (2) everytime you make a changes and want to build t
...
+- `release-publish`: Similar to `publish`, except the images will be uploaded with 'latest' and 'VERSION' tag. You can specify the image version in `VERSION` file.
```
[the remark this answers is not in the document]
Comment (marfgold1) [reply README.md]: I thought that the task needs to separate release and latest command, I misunderstood then? Okay I'll think about that

File: VERSION [hunk 16]
```diff
@@ -0,0 +1 @@
+0.1.0
```
[the remark this answers is not in the document]
Comment (marfgold1) [reply VERSION]: Oki doki :+1:

File: README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (marfgold1) [reply README.md]: Oh yeah I forgot to notice that clean will delete only images tagged with `DOCKER_USERNAME` too. That's why making `clean` to be universal need a shell script, because current implementation only purged images with `DOCKER_USERNAME` as registry and current image version as tag. The caveat is just it will also purge unintended images, for example if user have two images, their custom images (which they build with `make develop`) and openwisp latest image (that they pull the images by theirself), then user want to purge only image with their own username and they don't want to purge openwisp images, then universal `clean` can't do that job because it will delete both nevertheless. Tracking which one is comes from `make` is hard and probably will generate some unintended changes. If its safe to make clean to be universal, then I'll look forward to use shell script instead.

File: build/openwisp_base/Dockerfile (hunk 3 above)
Comment (marfgold1) [reply build/openwisp_base/Dockerfile]: I stilk have to change this right?

File: .travis.yml [hunk 17]
```diff
@@ -13,4 +13,9 @@ jobs:
   include:
     - stage: tests
       script:
-      - make travis-runtests
+      - |
+        if [ "$TRAVIS_PULL_REQUEST" != "false" ]; then
+          make travis-runtests
+        else
+          make travis-publish
```
Comment (atb00ker) [inline .travis.yml]: Also, please make sure that nightly publish only happens from the `master` branch! :smile:

File: README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (atb00ker) [reply README.md]: No, `latest` is only a special tag that is suppose to point to the latest stable version release! :smile:

File: README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (atb00ker) [reply README.md]: In my head, I think `clean` is very aggressive, and if someone is using it, I assume they want to remove everything, I think it's safe to assume if someone is using `clean` they just want to purge all the tags, the new, the old, everything.

Also, can you please mention that clean is an agressive `clean` this to the documentation?

We can have a soft clean in the future that take cares of the tags and what not in the future as per the requirements, but this `clean` came into existence for removing everything and I think we should stay true to that goal and focus on the rest later outside this issue.

What do you think? :smile:

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (atb00ker) [reply README.md]: Users will only need to deploy their own images when they want to change something in the image, i.e build the image.
I can't think of a senario where I would want a seperate image to manage for myself but I wouldn't want to change / build the images, can you share what senario you are thinking? :smile:

[the remark this answers is not in the document]
Comment (atb00ker) [reply README.md]: > That is important and if the users forgot, then docker-compose will use openwisp latest images instead. Also I do agree with that example, but then if you want to use multiple make calls, you have to export the variable instead of repeating yourself adding that DOCKER_USERNAME=*

Well, it's technically a requirement only for `make publish`, right? So we should mention it in relation to that only in my opinion, if someone has to use `make` multiple times without mentioning the environment variable, I guess they will have to learn environment variables! :smile:

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (atb00ker) [reply Makefile]: No, I just mean that it's harder to read this part, can use you spaces and `\` character to make part look clean, not just line `88` but couple of lines around it as well. :smile:

File: Makefile (hunk 8 above)
[the remark this answers is not in the document]
Comment (atb00ker) [reply Makefile]: I am not particular about it, I just think 12 factor asks us to keep management / one-off commands seperate.

I'll defer this to @2stacks or @nemesisdesign

File: Makefile (hunk 6 above)
[the remark this answers is not in the document]
Comment (atb00ker) [reply Makefile]: > Well dont blame me

Don't worry, nobody is blaming you for anything, even if you did it, reviews only exists to potentially improve the code not to blame someone on their coding styles.
Even if there is a mistake that you make, it's not an attack on you, blaming is not how organization can flurish.

Or in other words, `attack the faulty code not the developer`! :smile:

[the remark this answers is not in the document]
Comment (marfgold1) [reply Makefile]: Don't worry, I'm not really offended tho 😅

File: README.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (marfgold1) [reply README.md]: I see, Okay I will try my best to not using a shell script 😊

[the remark this answers is not in the document]
Comment (2stacks) [reply README.md]: When I think of `make clean` I mostly think of working with C code and I agree that it is usually there to restore the local repository to its original state.  That would mean purging everything created by make since cloning the repository.

File: Makefile (hunk 8 above)
[the remark this answers is not in the document]
Comment (2stacks) [reply Makefile]: I would expect the user to be logged in already.  I can't recall the last time I had to log in to Docker or Github.  It might be good that the execution stops as it would indicate that there's an issue with cached credentials.

File: README.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (2stacks) [reply README.md]: I think it should be mentioned at the first part of the workflow in which it will be needed.  If it's not needed until you actually push out images then I would not mention it early on.  I often build the containers to test development but never have a need to push them to dockerhub.

File: Makefile (hunk 7 above)
Comment (2stacks) [inline Makefile]: I know this wasn't added as part of this pull but isn't the redundant?  Seems like it should only call `develop-runtests`

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (2stacks) [reply Makefile]: Should `docker.pkg.github.com` be a variable?  Someone may want to use Gitlab or some other private registry.

File: Makefile (hunk 7 above)
Comment (marfgold1) [reply Makefile]: Yeah I'm not forgetting this part too, thanks for reminding :smile:

File: build/openwisp_base/Dockerfile (hunk 3 above)
Comment (atb00ker) [reply build/openwisp_base/Dockerfile]: Rebase to master! :smile:

**Comment by marfgold1** [comment]
I'll be on vacation tomorrow (don't worry it's just a short break). Oh and yeah, I change `DOCKER_USERNAME` to `REGISTRY` because I think it's more appropriate, though you still **have to** add both `REGISTRY` and `DOCKER_USERNAME` in the travis. :smile:
To-do:
1. Fix `make clean` to be universal
2. A little more update for `README.md`
You can review current commit so I can just read them later to save time, but it's still have missing part :smiley: 
Thanks guys

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (marfgold1) [reply Makefile]: @2stacks I don't think someone will be using `travis-publish`, besides `publish` is only supported for docker hub only :smile: 
The task isn't require me to add that functionality, but if you guys want me to add it, then I can do it but it'll take more time :sweat_smile:

**Comment by marfgold1** [comment]
I'm done :smile:
I hope it will resolve most of the reviews (I'm still not sure about readability because it's hard to make them short)

File: README.md [hunk 18]
```diff
@@ -172,6 +173,7 @@ The Makefile has following options, that are useful for developers & users. Defa
 - `runtests`: Run testcases to ensure all the services are working.
 - `develop-runtests`: Similar to `runtests`, it runs the testcases except doesn't stop the containers after running the tests which maybe desired for debugging & analyzing failing container's logs.
 - `travis-runtests`: Similar to `runtests`, it runs the testcases except used in travis builds because it doesn't clean up after running the tests and prints all the container's logs when an error occurs in the tests.
-- `clean`: Purge everything produced during building the images.
+- `clean`: Purge everything produced during building the images. **WARNING: This will clean ALL tagged \*/openwisp-\* images.**
```
Comment (atb00ker) [inline README.md]: Please change to:

```
Aggressively purge all the containers, images, volumes & networks related to `docker-openwisp`.
```

File: README.md [hunk 19]
```diff
@@ -172,6 +173,7 @@ The Makefile has following options, that are useful for developers & users. Defa
...
 - `develop`: Useful for development. Bundles all the commands required to build the images and run containers during development. With this option, you can interact with the interface and APIs and the logs will be printed on the terminal for debugging on the debugging level you select from the `.env` file.
-- `publish`: Build, test and publish the latest Official images.
+- `publish [REGISTRY=<your-image-registry>]`: Build, test and publish the latest official images to specified registry (default is `openwisp`). The images will be tagged as `VERSION` and `latest`. If you want to publish to another registry, be sure to use the correct format (For example, GitHub Packages use `docker.pkg.github.com/<github-username>/docker-openwisp` format).
+- `travis-publish`: Similar to `publish`, except used in travis builds because the images will be tagged as `nightly` and will be pushed to Docker Hub and GitHub Packages in an automated way.
```
Comment (atb00ker) [inline README.md]: - `travis-publish`: Similar to `publish`, except creates and pushes `nightly` images to Docker Hub and GitHub Packages.

File: Makefile (hunk 10 above)
[the remark this answers is not in the document]
Comment (2stacks) [reply Makefile]: >besides publish is only supported for docker hub only

That doesn't make sense to me.  You have a separate push command for docker hub and git hub.  I was asking if this can be reduced to a single command that can support dockerhub, github, gitlab and custom registries.  They are all using the docker api so a single command "should" work for all of them.

[the remark this answers is not in the document]
Comment (marfgold1) [reply Makefile]: Sorry, yeah now it's supported, I forgot to say in my current commit that this is possible, but I already did change the README so user will notice it :sweat_smile:

**Comment by atb00ker** [comment]
@nemesisdesign @2stacks what do you guys think about it?

Related issue: https://github.com/openwisp/docker-openwisp/issues/76

As disscussed on gitter with @2stacks. The automatic build of docker hub doesn't work for this repository because it only builds basic images and ignores the additional features added with docker-compose.

File: Makefile [hunk 20]
```diff
@@ -50,12 +53,34 @@ clean:
...
+	images=$$(docker-compose config | awk '{if($$0 ~ /image: $(REGISTRY)/ && !visited[$$0]++) print $$2}'); \
+	for image in $$images; do \
+		img_target=`echo $$image | sed "s/:$(TAG)/:latest/g"`; \
+		docker tag $$image $$img_target; \
+		docker push $$img_target; \
+	done
+
+travis-publish: TAG := nightly
+travis-publish: travis-runtests
+	echo "$(DOCKER_PASSWORD)" | docker login -u "$(DOCKER_USERNAME)" --password-stdin
```
Comment (atb00ker) [inline Makefile]: Just leaving a comment for myself : `docker rmi $(docker images |grep 'openwisp/')`

**Comment by atb00ker** [comment]
Hey marfgold, thanks for this work, I am still not sure If I would be able to use it, but I created a new branch `autoreleases` with your commit, I'll make the changes and use it from there if It seems like the right thing to do! :smile:
